### PR TITLE
sql: update generated column content

### PR DESCRIPTION
**Pull request by CaitinChen**
Via https://github.com/pingcap/docs-cn/pull/1283
@lilin90 @crazycs520 PTAL

### CODE REVIEW

File: sql/generated-columns.md [hunk 1]
```diff
@@ -8,7 +8,7 @@ category: user guide
 
 TiDB supports generated columns as part of MySQL 5.7 compatibility. One of the primary use cases for generated columns is to extract data out of a JSON data type and enable it to be indexed.
 
-## Index JSON using generated column
+## Index JSON using generated stored column
```
Comment (morgo) [inline sql/generated-columns.md]:
```suggestion
## Index JSON using stored generated column
```

File: sql/generated-columns.md [hunk 2]
```diff
@@ -21,21 +21,27 @@ CREATE TABLE person (
 );
 ```
 
-In order to index a JSON column, you must first extract it as a generated column. Using the `city` generated column as an example, you are then able to add an index:
+To index a JSON column, you must first extract it as a generated stored column. 
+
+> **Note:**
+>
+> The optimizer in TiDB only uses the index created on a generated stored column. Currently, the optimizer cannot use an index created on a generated virtual column. This issue will be fixed in the later TiDB versions (See ISSUE [#5189](https://github.com/pingcap/tidb/issues/5189)).
```
Comment (morgo) [inline sql/generated-columns.md]:
```suggestion
> The optimizer in TiDB only uses the index created on a generated stored column. Currently, the optimizer cannot use an index created on a generated virtual column. This issue will be fixed in the later TiDB versions (See [#5189](https://github.com/pingcap/tidb/issues/5189)).
```

File: sql/generated-columns.md [hunk 3]
```diff
@@ -21,21 +21,27 @@ CREATE TABLE person (
...
     KEY (city)
 );
 ```
 
-In this table, the `city` column is a **generated column**. As the name implies, the column is generated from other columns in the table, and cannot be assigned a value when inserted or updated. The column is also _virtual_ in that it does not require any storage or memory, and is generated on demand. The index on `city` however is _stored_ and uses the same structure as other indexes of the type `varchar(64)`.
+In this table, the `city` column is a **generated column**. As the name implies, the column is generated from other columns in the table, and cannot be assigned a value when inserted or updated. This column is generated based on the expression it defines and it is stored in the database. Thus this column can be read directly, not in a way that the `address_info` column it depends on is read first and then the data is calculated. The index on `city` however is _stored_ and uses the same structure as other indexes of the type `varchar(64)`.
 
-You can use the index on the generated column in order to speed up the following statement:
+You can use the index on the generated stored column in order to speed up the following statement:
```
Comment (morgo) [inline sql/generated-columns.md]:
```suggestion
You can use the index on the stored generated column in order to speed up the following statement:
```

File: sql/generated-columns.md [hunk 4]
```diff
@@ -60,6 +66,19 @@ mysql> INSERT INTO person (name, address_info) VALUES ('Morgan', JSON_OBJECT('Co
 ERROR 1048 (23000): Column 'city' cannot be null
 ```
 
+## Use generated virtual columns
+
+TiDB also supports generated virtual columns. Different from generated store columns, generated virtual columns are **virtual** in that they are generated as needed and are not stored in the database or occupied the memory space. Although TiDB supports indexing generated virtual columns, the optimizer currently cannot use the index in this case. This issue will be fixed in the later TiDB versions. (ISSUE [#5189](https://github.com/pingcap/tidb/issues/5189)).
```
Comment (morgo) [inline sql/generated-columns.md]:
```suggestion
TiDB also supports generated virtual columns. Different from generated store columns, generated virtual columns are **virtual** in that they are generated as needed and are not stored in the database or occupied the memory space. Although TiDB supports indexing generated virtual columns, the optimizer currently cannot use the index in this case. This issue will be fixed in a later version of TiDB ([#5189](https://github.com/pingcap/tidb/issues/5189)).
```

**Comment by CaitinChen** [comment]
@morgo Thanks~   @lilin90 @crazycs520 PTAL

File: sql/generated-columns.md [hunk 5]
```diff
@@ -21,21 +21,27 @@ CREATE TABLE person (
...
 ```
 
-In order to index a JSON column, you must first extract it as a generated column. Using the `city` generated column as an example, you are then able to add an index:
+To index a JSON column, you must first extract it as a generated stored column. 
+
+> **Note:**
+>
+> The optimizer in TiDB only uses the index created on a generated stored column. Currently, the optimizer cannot use an index created on a generated virtual column. This issue will be fixed in the later TiDB versions (See [#5189](https://github.com/pingcap/tidb/issues/5189)).
+
+Using the `city` generated stored column as an example, you are then able to add an index:
```
Comment (morgo) [inline sql/generated-columns.md]: Sorry, one more: 
```suggestion
Using the `city` stored generated column as an example, you are then able to add an index:
```

File: sql/generated-columns.md [hunk 6]
```diff
@@ -21,21 +21,27 @@ CREATE TABLE person (
 );
 ```
 
-In order to index a JSON column, you must first extract it as a generated column. Using the `city` generated column as an example, you are then able to add an index:
+To index a JSON column, you must first extract it as a generated stored column. 
+
+> **Note:**
+>
+> The optimizer in TiDB only uses the index created on a generated stored column. Currently, the optimizer cannot use an index created on a generated virtual column. This issue will be fixed in the later TiDB versions (See [#5189](https://github.com/pingcap/tidb/issues/5189)).
```
Comment (dcalvin) [inline sql/generated-columns.md]:
```suggestion
> The optimizer in TiDB only uses the index created on a generated stored column. Currently, the optimizer cannot use an index created on a generated virtual column. This issue will be fixed in later TiDB versions (See [#5189](https://github.com/pingcap/tidb/issues/5189)).
```

File: sql/generated-columns.md [hunk 7]
```diff
@@ -60,6 +66,19 @@ mysql> INSERT INTO person (name, address_info) VALUES ('Morgan', JSON_OBJECT('Co
 ERROR 1048 (23000): Column 'city' cannot be null
 ```
 
+## Use generated virtual columns
+
+TiDB also supports generated virtual columns. Different from generated store columns, generated virtual columns are **virtual** in that they are generated as needed and are not stored in the database or occupied the memory space. Although TiDB supports indexing generated virtual columns, the optimizer currently cannot use the index in this case. This issue will be fixed in a later version of TiDB ([#5189](https://github.com/pingcap/tidb/issues/5189)).
```
Comment (dcalvin) [inline sql/generated-columns.md]:
```suggestion
TiDB also supports generated virtual columns. Different from generated store columns, generated virtual columns are **virtual** in that they are generated as needed and are not stored in the database or cached in the memory. Although TiDB supports indexing generated virtual columns, the optimizer currently cannot use indexes in this case. This issue will be fixed in a later version of TiDB ([#5189](https://github.com/pingcap/tidb/issues/5189)).
```

File: sql/generated-columns.md [hunk 8]
```diff
@@ -21,21 +21,27 @@ CREATE TABLE person (
...
     address_info JSON,
-    city VARCHAR(64) AS (JSON_UNQUOTE(JSON_EXTRACT(address_info, '$.city'))) VIRTUAL,
+    city VARCHAR(64) AS (JSON_UNQUOTE(JSON_EXTRACT(address_info, '$.city'))) STORED,
     KEY (city)
 );
 ```
 
-In this table, the `city` column is a **generated column**. As the name implies, the column is generated from other columns in the table, and cannot be assigned a value when inserted or updated. The column is also _virtual_ in that it does not require any storage or memory, and is generated on demand. The index on `city` however is _stored_ and uses the same structure as other indexes of the type `varchar(64)`.
+In this table, the `city` column is a **generated column**. As the name implies, the column is generated from other columns in the table, and cannot be assigned a value when inserted or updated. This column is generated based on the expression it defines and it is stored in the database. Thus this column can be read directly, not in a way that the `address_info` column it depends on is read first and then the data is calculated. The index on `city` however is _stored_ and uses the same structure as other indexes of the type `varchar(64)`.
```
Comment (dcalvin) [inline sql/generated-columns.md]:
```suggestion
In this table, the `city` column is a **generated column**. As the name implies, the column is generated from other columns in the table, and cannot be assigned a value when inserted or updated. This column is generated based on a defined expression and is stored in the database. Thus this column can be read directly, not in a way that its dependent column `address_info`  is read first and then the data is calculated. The index on `city` however is _stored_ and uses the same structure as other indexes of the type `varchar(64)`.
```

**Comment by CaitinChen** [comment]
@crazycs520 PTAL

**Comment by CaitinChen** [comment]
Because conflicts occur, I close this PR and file another: https://github.com/pingcap/docs/pull/1232

**Pull request closed without merging**